Feat: Add pre and post update, start, backup, and shutdown discord hooks.

v4de opened the pull request:
## Context <!-- markdownlint-disable MD041 -->

* I would like to be able to add custom sh commands and discord webhooks before and after server is started or shutdown. Backup hooks were added as a bonus because I felt it may be useful in the future.

## Choices

* Used [lloesche/valheim-server-docker](https://github.com/lloesche/valheim-server-docker) previously and used this feature while hosting the container in the past. Would like to have the same functionality or better

## Test instructions

### Tested discord webhooks

1. `docker build . -t v4de/palworld-server-docker:0.1`
2. Generate a webhook to be for with discord
3. Added following environment variables to docker-compose.yml

```yaml
  - DISCORD_WEBHOOK=<place-webhook-here>
  - DISCORD_PRE_INIT_MESSAGE="Server initializing..."
  - DISCORD_POST_INIT_MESSAGE="Server started"
  - DISCORD_PRE_BACKUP_MESSAGE="Trying to backup..."
  - DISCORD_POST_BACKUP_MESSAGE="Backup created at $$FILE_PATH"
  - DISCORD_PRE_SHUTDOWN_MESSAGE="Server shutting down..."
  - DISCORD_POST_SHUTDOWN_MESSAGE="Server stopped"
  - 'PRE_INIT_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_PRE_INIT_MESSAGE)\"}" $$DISCORD_WEBHOOK'
  - 'POST_INIT_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_POST_INIT_MESSAGE)\"}" $$DISCORD_WEBHOOK'
  - 'PRE_BACKUP_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_PRE_BACKUP_MESSAGE)\"}" $$DISCORD_WEBHOOK'
  - 'POST_BACKUP_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_POST_BACKUP_MESSAGE)\"}" $$DISCORD_WEBHOOK'
  - 'PRE_SHUTDOWN_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_PRE_SHUTDOWN_MESSAGE)\"}" $$DISCORD_WEBHOOK'
  - 'POST_SHUTDOWN_HOOK=curl -sfSL -H "Content-Type: application/json" -d "{\"username\":\"Palworld\",\"content\":\"$$(eval echo $$DISCORD_POST_SHUTDOWN_MESSAGE)\"}" $$DISCORD_WEBHOOK'
```

4. `docker compose up -d --no-deps --build`
5. `docker exec palworld-server backup`
6. `docker compose down`
7. Validate expected messages received by discord
![image](https://github.com/thijsvanloef/palworld-server-docker/assets/934222/42d98736-6f69-4dbb-905f-b902f0890552)

### Tested backup after shutdown

1. Added following environment variable

```yaml
  - POST_SHUTDOWN_HOOK=backup
```

2. `docker compose up -d`
3. note current backups and timestamps
4. `docker compose down`
5. found additional backup has been created with latest timestamp.

## Checklist before requesting a review

* [x] I have performed a self-review of my code
* [x] I've added documentation about this change to the README.
* [x] I've not introduced breaking changes.

Dashboy1998 reviewed (commented):
Not a good idea to run `eval $var` in init.sh as that will be ran as root.

I don't recommend running the anything before the shutdown command. If for some reason `eval $PRE_SHUTDOWN_HOOK` is frozen then your server will not save.

I recommend moving the post shutdown statement to the end of start.sh. Not sure how you could safety run the pre shutdown safely.

v4de commented:
Well these are optional and would allow others to extend the container as they want. I understand that this should be only for advanced users, maybe I can make a note of that in the readme.

Unfortunately palworld is currently not running as a service so its difficult for me to grab an accurate status. This could be another enhancement. 

When I place the hook at the end of start.sh I am unable to send a message && sleep 60 seconds, as the term_handler will have already run the rcon shutdown before anyone is notified.

```console
2024-01-28T19:35:29.369379798Z *****SIGTERM BEGIN*****
2024-01-28T19:35:29.395356988Z Complete Save
2024-01-28T19:35:29.425128085Z The server will shut down in 1 seconds. Please prepare to exit the game.
2024-01-28T19:35:33.034047136Z dlmopen steamservice.so failed: steamservice.so: cannot open shared object file: No such file or directory
2024-01-28T19:35:33.852180758Z Shutdown handler: cleanup.
2024-01-28T19:35:33.875764572Z *****HI START.SH END*****
2024-01-28T19:35:34.427063991Z *****SIGTERM END*****
2024-01-28T19:35:34.427682703Z *****HI INIT.SH END*****
```

If the concern is security I could make this only for the discord hooks for now, as this is the most valuable for me right now:

```sh
DISCORD_WEBHOOK="https://discord.com/api/webhooks/$DISCORD_WEBHOOK_ID"

  if [ -n "$DISCORD_WEBHOOK" ]; then
      echo "DISCORD_PRE_INIT_JSON: $DISCORD_PRE_INIT_JSON"
      curl -sfSL -H "Content-Type: application/json" -d "$DISCORD_PRE_INIT_JSON" "$DISCORD_WEBHOOK"'
  fi
```

I can also move the hooks to another sh and have it run by steam user only:
```sh 
# run as steam user
su steam -c ./hook.sh discord_hook $DISCORD_JSON $DISCORD_WEBHOOK &
```

Open to more suggestions as well.

thijsvanloef commented:
tbh, I really like the dedicated discord hooks idea, keeps it more simple and I can imagine that a lot of people would like such a feature

Dashboy1998 commented:
I think a Discord notifier is a great idea and I'm eager to use it myself.

Term is meant to catch stopping the container so it must shutdown quickly. It's about preserving the save data.
For shutdown I would create a separate shutdown script and have that run the notifier. Although to get notified you'd have to run `docker exec palworld-server nice_shutdown` but that way you could set the timeout instead of 1 second.

Recommend timeout for curl: `curl --connect-timeout $DISCORD_TIMEOUT ...`

If you're only doing discord then you need at least these environmental vars
* DISCORD_WEBHOOK
* DISCORD_USER
* DISCORD_TIMEOUT
* NICE_SHUTDOWN_TIME

The messages as environmental variables are a nice touch although I'm not sure how necessary they are.

thijsvanloef commented:
@v4de is this PR still active?

v4de commented:
hey @thijsvanloef, ill work on this today. I have been busy with other work.

thijsvanloef commented:
@v4de no problem at all! Was just curious, take your time :)

v4de commented:
> The messages as environmental variables are a nice touch although I'm not sure how necessary they are.

@Dashboy1998 We could make some standard json formatting which will reduce the amount of variables. Currently I have been using a dansgaming lamball icon and some basic colors and messages:

![image](https://github.com/thijsvanloef/palworld-server-docker/assets/934222/59f1030a-01f0-41ae-8a51-2d8d72759c44)

Not sure how to give the ability to choose to have some messages and not others. Will just add to all the places I believe helpful for notifications right now.

Dashboy1998 commented:
> > The messages as environmental variables are a nice touch although I'm not sure how necessary they are.
> 
> @Dashboy1998 We could make some standard json formatting which will reduce the amount of variables. Currently I have been using a dansgaming lamball icon and some basic colors and messages:
> 
> ![image](https://private-user-images.githubusercontent.com/934222/301232722-59f1030a-01f0-41ae-8a51-2d8d72759c44.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3MzA4ODAsIm5iZiI6MTcwNjczMDU4MCwicGF0aCI6Ii85MzQyMjIvMzAxMjMyNzIyLTU5ZjEwMzBhLTAxZjAtNDFhZS04YTUxLTJkOGQ3Mjc1OWM0NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEzMVQxOTQ5NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZWU5YmQwOTZmOTgyNThkYjQ0YTlmMzEzZWUxNDNjNjhiM2MyYmE5NzM2YTQzZjQ1M2RjYzZjMTUyOWMxYWVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Hy5tCDUVf-EfRpkLXyWzPj23R3K_uPTLbaKLclRQBBs)
> 
> Not sure how to give the ability to choose to have some messages and not others. Will just add to all the places I believe helpful for notifications right now.

I think the message variables are a good idea although I would specify default values for the messages, timeout, and even user.

Still not sure how to get a good shutdown warning other than a custom shutdown script otherwise I worry about people changing the NICE_SHUTDOWN_TIME but not the stop_grace_period.

thijsvanloef commented:
Hi @v4de, Thanks for the work! will look at it tonight!

Dashboy1998 commented on README.md:
```diff
@@ -184,6 +184,14 @@ It is highly recommended you set the following environment values before startin
 | AUTO_REBOOT_CRON_EXPRESSION  | Setting affects frequency of automatic updates. | 0 0 \* \* \* | Needs a Cron-Expression - See [Configuring Automatic Backups with Cron](#configuring-automatic-reboots-with-cron) |
 | AUTO_REBOOT_ENABLED | Enables automatic reboots | false | true/false |
 | AUTO_REBOOT_WARN_MINUTES | How long to wait to reboot the server, after the player were informed. | 5 | !0 |
+| DISCORD_WEBHOOK_ID | Discord webhook id found after creating a webhook on a discord server ex: discord.com/api/webhooks/<webhook_id> | 0123456789012345678/xxxxxxxxxxxxxxxxxxxxxx | |
```
I would change this to be the entire URL `https://discord.com/api/webhooks/xxxxxxxxxxxxxxxxxxxx/abcdef`. Makes it easier for users.

Also there is no default value.

v4de replied:
Thats an ez change, when I put the url in the readme tho I get a linting error for using a bare url. Anyway to get around that or we should ignore it?

About the default value, there should not be one as the value is unique and must be generated.

Dashboy1998 replied:
Right there should be a default value. It looked to me like there was a default value listed in the Readme.

Dashboy1998 commented on Dockerfile:
```diff
@@ -1,16 +1,17 @@
-      dockerhub="https://hub.docker.com/r/thijsvanloef/palworld-server-docker" \
-      org.opencontainers.image.authors="Thijs van Loef" \
-      org.opencontainers.image.source="https://github.com/thijsvanloef/palworld-server-docker"
+    name="thijsvanloef/palworld-server-docker" \
+    github="https://github.com/thijsvanloef/palworld-server-docker" \
+    dockerhub="https://hub.docker.com/r/thijsvanloef/palworld-server-docker" \
+    org.opencontainers.image.authors="Thijs van Loef" \
+    org.opencontainers.image.source="https://github.com/thijsvanloef/palworld-server-docker"
```
```suggestion
      name="thijsvanloef/palworld-server-docker" \
      github="https://github.com/thijsvanloef/palworld-server-docker" \
      dockerhub="https://hub.docker.com/r/thijsvanloef/palworld-server-docker" \
      org.opencontainers.image.authors="Thijs van Loef" \
      org.opencontainers.image.source="https://github.com/thijsvanloef/palworld-server-docker"
```

While your changes here are meant to meet the 4 spaces I said earlier, the purpose for these extra spaces are so it's more easily readable as a continuation of LABEL. 
I would change this back so you're not listed in git blame.

v4de replied:
Understood, it was just to get my linter to stop yelling at me, can be changed back.

Dashboy1998 commented on Dockerfile:
```diff
@@ -27,16 +28,16 @@ ENV RCON_MD5SUM="8601c70dcab2f90cd842c127f700e398" \
-     && rm rcon.tar.gz \
-     && mv rcon-${RCON_VERSION}-amd64_linux/rcon /usr/bin/rcon-cli \
-     && rmdir /tmp/dumps
+    && echo "${RCON_MD5SUM}" rcon.tar.gz | md5sum -c - \
+    && tar -xzvf rcon.tar.gz \
+    && rm rcon.tar.gz \
+    && mv rcon-${RCON_VERSION}-amd64_linux/rcon /usr/bin/rcon-cli \
+    && rmdir /tmp/dumps
```
I have no idea if this should be changed or not. I'll let @thijsvanloef decide.

Dashboy1998 commented on Dockerfile:
```diff
@@ -27,16 +28,16 @@ ENV RCON_MD5SUM="8601c70dcab2f90cd842c127f700e398" \
-     && echo "${SUPERCRONIC_SHA1SUM}" supercronic | sha1sum -c - \
-     && chmod +x supercronic \
-     && mv supercronic /usr/local/bin/supercronic
+    && echo "${SUPERCRONIC_SHA1SUM}" supercronic | sha1sum -c - \
+    && chmod +x supercronic \
+    && mv supercronic /usr/local/bin/supercronic
```
I have no idea if this should be changed or not. I'll let @thijsvanloef decide.

Dashboy1998 commented on Dockerfile:
```diff
@@ -64,14 +65,22 @@ ENV PORT= \
     mv /home/steam/server/update.sh /usr/local/bin/update && \
-    mv /home/steam/server/restore.sh /usr/local/bin/restore
+    mv /home/steam/server/restore.sh /usr/local/bin/restore && \
+    mv /home/steam/server/discord.sh /usr/local/bin/discord
```
```suggestion
    mv /home/steam/server/discord.sh /usr/local/bin/discord
```

I don't think we should make discord a command. I would recommend calling the script directly.

v4de replied:
good idea, makes sense

Dashboy1998 commented on docker-compose.yml:
```diff
@@ -1,25 +1,25 @@
-         - SERVER_DESCRIPTION=palworld-server-docker by Thijs van Loef
-      volumes:
-         - ./palworld:/palworld/
+  palworld:
+    image: thijsvanloef/palworld-server-docker:latest
+    restart: unless-stopped
+    container_name: palworld-server
+    stop_grace_period: 30s # Set to however long you are willing to wait for the container to gracefully stop
+    ports:
+      - 8211:8211/udp
+      - 27015:27015/udp # Required if you want your server to show up in the community servers tab
+    environment:
+      - PUID=1000
+      - PGID=1000
+      - PORT=8211 # Optional but recommended
+      - PLAYERS=16 # Optional but recommended
+      - SERVER_PASSWORD=worldofpals # Optional but recommended
+      - MULTITHREADING=true
+      - RCON_ENABLED=true
+      - RCON_PORT=25575
+      - TZ=UTC
+      - ADMIN_PASSWORD=adminPasswordHere
+      - COMMUNITY=false  # Enable this if you want your server to show up in the community servers tab, USE WITH SERVER_PASSWORD!
+      - SERVER_NAME=World of Pals
+      - SERVER_DESCRIPTION=palworld-server-docker by Thijs van Loef
+    volumes:
+      - ./palworld:/palworld/
```
We currently use 3 spaces in this YAML only. 
I do support this update to 2 spaces since all the other YAMLs are 2.

Dashboy1998 commented on scripts/init.sh:
```diff
@@ -1,11 +1,11 @@
 
 if [[ ! "${PUID}" -eq 0 ]] && [[ ! "${PGID}" -eq 0 ]]; then
-    printf "\e[0;32m*****EXECUTING USERMOD*****\e[0m\n"
+    printf "\e[0;32m%s\e[0m\n" "*****EXECUTING USERMOD*****"
```
What is the purpose of changing the printf(s)?

Dashboy1998 commented on scripts/start.sh:
```diff
@@ -125,7 +134,7 @@ if [ -n "${SERVER_PASSWORD}" ]; then
 if [ -n "${ADMIN_PASSWORD}" ]; then
     ADMIN_PASSWORD=$(sed -e 's/^"\(.*\)"$/\1/' -e "s/^'\(.*\)'$/\1/" <<< "$ADMIN_PASSWORD")
-    echo "ADMIN_PASSWORD=${ADMIN_PASSWORD}" 
+    echo "ADMIN_PASSWORD=${ADMIN_PASSWORD}"
```
Thank you for removing the extra space.

Dashboy1998 reviewed (changes requested):
![image](https://github.com/thijsvanloef/palworld-server-docker/assets/18587674/ab9a4696-a2b6-4503-ad2a-c0593f887a10)

Assuming your game is update to date then the order update completed message may be displayed before the server is started message.

If there is an update then the order is correct. It's just a timing issue in the script.

I am getting no messages for running backup.
I am getting no messages for running update.

Please fix all the spell check errors for shell SC2086.

v4de commented:
@thijsvanloef Changed the syntax for the discord.sh to make it less messy. I wanted to expose the script some how, but agree no good way to do that right now. Also agree with moving it out of usr/bin. Sorry for all the linting changes, will try to keep that very minimal in the future.

v4de commented:
Replying to @Dashboy1998 :
> Assuming your game is update to date then the order update completed message may be displayed before the server is started message.
> If there is an update then the order is correct. It's just a timing issue in the script.

I have seen this also in my testing, I am not sure why it was happening so quick, I removed the success from running in the background anymore.
 
> I am getting no messages for running backup. I am getting no messages for running update.

This is interesting, I was seeing backup messages in my testing. Let me test this again, maybe I changed something after I tested.

> Please fix all the spell check errors for shell SC2086.

I think @thijsvanloef assisted with the linting, I also updated some when changing the command.

Dashboy1998 commented:
> > If there is an update then the order is correct. It's just a timing issue in the script.
> 
> I have seen this also in my testing, I am not sure why it was happening so quick, I removed the success from running in the background anymore.

I suspect the issue is one curl command beating the other. Not sure how to prevent that other than not running it in the background like you changed it or going to the complicated process of making a FIFO for the discord messages. 


> > I am getting no messages for running backup. I am getting no messages for running update.
> 
> This is interesting, I was seeing backup messages in my testing. Let me test this again, maybe I changed something after I tested.

Strange. Here's my docker compose

```YAML
services:
   palworld:
      image: thijsvanloef/palworld-server-docker:test
      build:
         dockerfile: Dockerfile
      restart: unless-stopped
      container_name: palworld-server-test
      stop_grace_period: 30s # Set to however long you are willing to wait for the container to gracefully stop
      ports:
        - 8212:8211/udp
        - 27015:27015/udp # Required if you want your server to show up in the community servers tab
      environment:
         - PUID=1000
         - PGID=1000
         - PORT=8211 # Optional but recommended
         - PLAYERS=16 # Optional but recommended
         - SERVER_PASSWORD=worldofpals # Optional but recommended
         - MULTITHREADING=true
         - RCON_ENABLED=true
         - RCON_PORT=25575
         - TZ=UTC
         - ADMIN_PASSWORD=adminPasswordHere
         - COMMUNITY=false  # Enable this if you want your server to show up in the community servers tab, USE WITH SERVER_PASSWORD!
         - SERVER_NAME=World of Pals
         - SERVER_DESCRIPTION=palworld-server-docker by Thijs van Loef
         - DISCORD_WEBHOOK_ID=1234/abde
      volumes:
         - ./palworld:/palworld/ 

```

v4de commented on README.md:
```diff
@@ -184,6 +184,14 @@ It is highly recommended you set the following environment values before startin
 | AUTO_REBOOT_CRON_EXPRESSION  | Setting affects frequency of automatic updates. | 0 0 \* \* \* | Needs a Cron-Expression - See [Configuring Automatic Backups with Cron](#configuring-automatic-reboots-with-cron) |
 | AUTO_REBOOT_ENABLED | Enables automatic reboots | false | true/false |
 | AUTO_REBOOT_WARN_MINUTES | How long to wait to reboot the server, after the player were informed. | 5 | !0 |
+| DISCORD_WEBHOOK_URL | Discord webhook url found after creating a webhook on a discord server | https://discord.com/api/webhooks/<webhook_id> | |
```
awesome that fixed it!

v4de commented:
Ok its working for me now, all the conditions were still checking for ID. Pull latest and use `DISCORD_WEBHOOK_URL` now instead of `DISCORD_WEBHOOK_ID` in the env variables. Changed per recommendation to make it easier for [users](https://github.com/thijsvanloef/palworld-server-docker/pull/172#discussion_r1478614397). I can change it back also if you prefer not to have url in the vars, not sure if I misunderstood the suggestion was only for readme or not.

![image](https://github.com/thijsvanloef/palworld-server-docker/assets/934222/d1caf445-7c09-4228-8fa0-3e1e937e0a3a)

Saving Backup:

```
PS D:\repo\palworld-server-docker> docker exec -it palworld-server backup
Sending Discord json: {"embeds":[{"title":"Creating backup...","color":15258703}]}
Complete Save
Creating backup
Backup created at /palworld/backups/palworld-save-2024-02-05_21-30-13.tar.gz
Sending Discord json: {"embeds":[{"title":"Backup created at /palworld/backups/palworld-save-2024-02-05_21-30-13.tar.gz","color":52224}]}
```

I do not have an update, so was not able to test updates, I can add a message for already up to date, but that might be too spammy with autoupdate crons now being a thing.

```
PS D:\repo\palworld-server-docker> docker exec -it palworld-server update
The Server is up to date!
```

Dashboy1998 commented:
> Ok its working for me now, all the conditions were still checking for ID. Pull latest and use `DISCORD_WEBHOOK_URL` now instead of `DISCORD_WEBHOOK_ID` in the env variables. Changed per recommendation to make it easier for [users](https://github.com/thijsvanloef/palworld-server-docker/pull/172#discussion_r1478614397). I can change it back also if you prefer not to have url in the vars, not sure if I misunderstood the suggestion was only for readme or not.
> 
> ![image](https://github.com/thijsvanloef/palworld-server-docker/assets/934222/d1caf445-7c09-4228-8fa0-3e1e937e0a3a)
> 
> Saving Backup:
> 
> ```
> PS D:\repo\palworld-server-docker> docker exec -it palworld-server backup
> Sending Discord json: {"embeds":[{"title":"Creating backup...","color":15258703}]}
> Complete Save
> Creating backup
> Backup created at /palworld/backups/palworld-save-2024-02-05_21-30-13.tar.gz
> Sending Discord json: {"embeds":[{"title":"Backup created at /palworld/backups/palworld-save-2024-02-05_21-30-13.tar.gz","color":52224}]}
> ```
> 
> I do not have an update, so was not able to test updates, I can add a message for already up to date, but that might be too spammy with autoupdate crons now being a thing.
> 
> ```
> PS D:\repo\palworld-server-docker> docker exec -it palworld-server update
> The Server is up to date!
> ```

I agree repeatedly stating an update is available is excessive. Take a look at #209 to see how to trick the update script into updating.

v4de commented:
> I agree repeatedly stating an update is available is excessive. Take a look at #209 to see how to trick the update script into updating.

Cool that worked and got the update warning!

![image](https://github.com/thijsvanloef/palworld-server-docker/assets/934222/d2748abf-97be-4bec-a60e-6fd51bd1e3da)

Dashboy1998 reviewed (approved):
Looks good to me.

thijsvanloef reviewed (approved):
made a small change in the text, but LGTM
